src/tls.c

```
@@ -1076,41 +1088,124 @@ sds connTLSGetPeerCert(connection *conn_) {
-}
+    case CTRL_TLS_GET_PEER_CERT:
+        *(sds *)arg1 = connTLSGetPeerCert((connection *)arg0);
+        return C_OK;
 
-int tlsConfigure(redisTLSContextConfig *ctx_config) {
-    UNUSED(ctx_config);
-    return C_OK;
-}
+    case CTRL_TLS_GET_CTX:
+        *(SSL_CTX **)arg0 = redis_tls_ctx;
+        return C_OK;
 
-connection *connCreateTLS(void) { 
-    return NULL;
+    case CTRL_TLS_GET_CLIENT_CTX:
+        *(SSL_CTX **)arg0 = redis_tls_client_ctx;
+        return C_OK;
+    }
+
+    serverLog(LL_WARNING, "TLS: Unknown connection control %ld", cmd);
+    return C_ERR;
 }
 
-connection *connCreateAcceptedTLS(int fd, int require_auth) {
-    UNUSED(fd);
-    UNUSED(require_auth);
+static ConnectionType CT_TLS = {
+    /* connection type */
+    .get_type = connTLSGetType,
 
-    return NULL;
-}
+    /* connection type initialize & finalize & configure */
+    .init = tlsInit,
+    .cleanup = tlsCleanup,
 
-int tlsHasPendingData() {
-    return 0;
+    /* ae & accept & listen & error & address handler */
+    .ae_handler = tlsEventHandler,
+    .accept_handler = tlsAcceptHandler,
+    .addr = connTLSAddr,
+    .listen = connTLSListen,
+
+    /* create/close connection */
+    .conn_create = connCreateTLS,
+    .conn_create_accepted = connCreateAcceptedTLS,
+    .close = connTLSClose,
+
+    /* connect & accept */
+    .connect = connTLSConnect,
+    .blocking_connect = connTLSBlockingConnect,
+    .accept = connTLSAccept,
+
+    /* IO */
+    .read = connTLSRead,
+    .write = connTLSWrite,
+    .writev = connTLSWritev,
+    .set_write_handler = connTLSSetWriteHandler,
+    .set_read_handler = connTLSSetReadHandler,
+    .get_last_error = connTLSGetLastError,
+    .sync_write = connTLSSyncWrite,
+    .sync_read = connTLSSyncRead,
+    .sync_readline = connTLSSyncReadLine,
+
+    /* pending data */
+    .has_pending_data = tlsHasPendingData,
+    .process_pending_data = tlsProcessPendingData,
+
+    /* TLS specified methods */
+    .control = tlsControl
```

This feels awkward. Why not have a separate TLS type here instead of a generic poorly typed control method. These still have to be called from Redis. something like:
```
tlsControl = {
    .get_certificate = whatever
    .get_peer = whatever
}

.tls_control = tlsControl
```

---

> This feels awkward. Why not have a separate TLS type here instead of a generic poorly typed control method. These still have to be called from Redis. something like:
> 
> ```
> tlsControl = {
>     .get_certificate = whatever
>     .get_peer = whatever
> }
> 
> .tls_control = tlsControl
> ```

From this design, we need define something in connection.h:
```
struct tlsControl = {
    sds (*get_peer_cert)(struct connection *conn);
    void* (*get_ctx)(void);
    void* (*get_client_ctx)(void);
    int (*configure)(void *priv, int reconfigure);
};

struct ConnectionType {
     ...
     struct tlsControl *tlsctrl;
};

static inline void *connTypeGetClientCtx(ConnectionType *ct) {
    return ct->tlsctrl->get_client_ctx();
}
...
```

My concern is:
* **struct ConnectionType** is a common/base class of every connection type, it should contain the common methods only.
* if we need introduce any other connection specific method, we need modify **struct ConnectionType**. (Ex, struct fooControl *fooctrl).

---

it was discussed here https://github.com/redis/redis/pull/9320#discussion_r912445302
and IIRC we concluded a generic `control` is better than polluting the interface with TLS specific concerns.

---

I don't really agree with the concern of "polluting the interface". The connection implementation is still tightly coupled to the core, since we explicitly call the TLS functions with specific arguments. If the interface changes, I presume that calling sites would rather have a type mismatch to know something has changed instead of silently ignoring or mishandling it.

---

> 
Hi @madolson 

Connection module is allowed only when the building version matches with Redis server.
This is quite similar to Linux kernel&modules.

---

> Connection module is allowed only when the building version matches with Redis server.
This is quite similar to Linux kernel&modules.

Right. So it seems like the added benefit of having verifying type is a good thing?

---

As it's an internal interface, I'm OK with both ways.

---

I'll also make a point since this is an internal interface, it's an easily reversible decision. The use case I'm imagining is that we wan to change the interface, but only update one of many calling sites. Since the compiler isn't validating the interface, we don't see an error and it fails at runtime.

---

ok, let's go back on this and re-add the specific (type safe) interfaces.
IIRC it's just 2 interfaces now since the other 2 were converted to a hack using global externs (outside of this interface)